Port smooth_l1_loss to structured kernels

alanwaketan opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* __->__ #67404

Summary:
Port smooth_l1_loss to structured kernels.

Brian Hirsh authored the part of adding build_borrowing_binary_op_coerce_to_scalar to TensorIterator.

Test Plan:
This commit shouldn't change the behavior. So, CI.

Differential Revision: [D31981147](https://our.internmc.facebook.com/intern/diff/D31981147)

Commit:
Port smooth_l1_loss to structured kernels

Summary:
Port smooth_l1_loss to structured kernels.

Test Plan:
This commit shouldn't change the behavior. So, CI.

[ghstack-poisoned]

Commit:
Update on "Port smooth_l1_loss to structured kernels"


Summary:
Port smooth_l1_loss to structured kernels.

Test Plan:
This commit shouldn't change the behavior. So, CI.

Differential Revision: [D31981147](https://our.internmc.facebook.com/intern/diff/D31981147)

[ghstack-poisoned]

Commit:
Update on "Port smooth_l1_loss to structured kernels"


Summary:
Port smooth_l1_loss to structured kernels.

Test Plan:
This commit shouldn't change the behavior. So, CI.

Differential Revision: [D31981147](https://our.internmc.facebook.com/intern/diff/D31981147)

[ghstack-poisoned]

Commit:
Update on "Port smooth_l1_loss to structured kernels"


Summary:
Port smooth_l1_loss to structured kernels.

Test Plan:
This commit shouldn't change the behavior. So, CI.

Differential Revision: [D31981147](https://our.internmc.facebook.com/intern/diff/D31981147)

[ghstack-poisoned]

Commit:
Update on "Port smooth_l1_loss to structured kernels"


Summary:
Port smooth_l1_loss to structured kernels.

Test Plan:
This commit shouldn't change the behavior. So, CI.

Differential Revision: [D31981147](https://our.internmc.facebook.com/intern/diff/D31981147)

[ghstack-poisoned]

Commit:
Update on "Port smooth_l1_loss to structured kernels"


Summary:
Port smooth_l1_loss to structured kernels.

Test Plan:
This commit shouldn't change the behavior. So, CI.

Differential Revision: [D31981147](https://our.internmc.facebook.com/intern/diff/D31981147)

[ghstack-poisoned]

Commit:
Update on "Port smooth_l1_loss to structured kernels"


Summary:
Port smooth_l1_loss to structured kernels.

Test Plan:
This commit shouldn't change the behavior. So, CI.

Differential Revision: [D31981147](https://our.internmc.facebook.com/intern/diff/D31981147)

[ghstack-poisoned]

Commit:
Update on "Port smooth_l1_loss to structured kernels"


Summary:
Port smooth_l1_loss to structured kernels.

Test Plan:
This commit shouldn't change the behavior. So, CI.

Differential Revision: [D31981147](https://our.internmc.facebook.com/intern/diff/D31981147)

[ghstack-poisoned]

alanwaketan commented on aten/src/ATen/native/Loss.cpp:
```diff
@@ -20,7 +22,24 @@ namespace {
+
+  TORCH_INTERNAL_ASSERT(reduction == Reduction::Mean || reduction == Reduction::Sum);
+  MetaBase::set_output({}, input.options());
+}
```
Sure, let me change to that.

alanwaketan replied:
``at::native::resize_output(Tensor, IntArrayRef)` seems deprecated.

bdhirsh commented on aten/src/ATen/native/Loss.cpp:
```diff
@@ -351,38 +391,6 @@ Tensor soft_margin_loss(
-    auto iter = TensorIterator::borrowing_binary_op(loss, input, target);
-    smooth_l1_stub(iter.device_type(), iter, beta);
-    if (reduction == Reduction::Mean) {
-      at::mean_out(result, iter.output(), 0);
```
yeah I think you're right, this was just broken. Tested:
```
>>> a = torch.ones(2, 2)
>>> b1 = torch.ones(())
>>> b1.size()
torch.Size([])
>>> torch._C._nn.smooth_l1_loss(a, a, beta=1.5, reduction=1, out=b1)
tensor([0., 0.])
```

It probably isn't hit anywhere by most people (or in tests) because most people use the loss function through the `torch.nn.functional()` API, which doesn't call the `out=` variant.

alanwaketan commented on aten/src/ATen/native/Loss.cpp:
```diff
@@ -29,6 +49,27 @@ DEFINE_DISPATCH(huber_backward_stub);
+    }
+  } else {
+    auto iter = TensorIterator::borrowing_binary_op(result, input, target);
+    smooth_l1_stub(iter.device_type(), iter, beta);
```
Figured this out, I need to change the signatures in as well:
```
REGISTER_DISPATCH(smooth_l1_stub, &smooth_l1_kernel);
REGISTER_DISPATCH(smooth_l1_stub, &smooth_l1_kernel_cuda);
```

alanwaketan commented on aten/src/ATen/native/Loss.cpp:
```diff
@@ -29,6 +48,27 @@ DEFINE_DISPATCH(huber_backward_stub);
+    }
+  } else {
+    TensorIterator iter = *this;
+    smooth_l1_stub(device_type(), iter, beta);
```
Done.

alanwaketan commented on aten/src/ATen/native/Loss.cpp:
```diff
@@ -29,6 +48,27 @@ DEFINE_DISPATCH(huber_backward_stub);
+  }
+  if (reduction != Reduction::None) {
+    Tensor loss;
+    auto iter = TensorIterator::borrowing_binary_op(loss, input, target);
```
Not sure if I follow. May be related to your question below for @ngimel.

Commit:
Update on "Port smooth_l1_loss to structured kernels"


Summary:
Port smooth_l1_loss to structured kernels.

Test Plan:
This commit shouldn't change the behavior. So, CI.

Differential Revision: [D31981147](https://our.internmc.facebook.com/intern/diff/D31981147)

[ghstack-poisoned]

alanwaketan commented:
@alanwaketan has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D31981147).

bdhirsh commented on aten/src/ATen/native/Loss.cpp:
```diff
@@ -20,7 +22,24 @@ namespace {
+TORCH_META_FUNC(smooth_l1_loss)
+(const Tensor& input, const Tensor& target, const int64_t reduction, double beta) {
+  TORCH_CHECK(beta >= 0, "smooth_l1_loss does not support negative values for beta.")
+  build_borrowing_binary_op(maybe_get_output(), input, target);
```
Hey @alanwaketan, here's a patch to do shape checking with TensorIterator, without needing to allocate output. I tried it locally on your PR and ran `pytest test/test_ops.py -k smooth_l1_loss` with no failures; hopefully this fixes the errors you were seeing?

```
z(/scratch/hirsheybar/pytorch-amin_symbol-env) hirsheybar@devfair017:/scratch/hirsheybar/pytorch-amin_symbol$ git diff
diff --git a/aten/src/ATen/TensorIterator.cpp b/aten/src/ATen/TensorIterator.cpp
index 6fe00accf1..9d6853eaa3 100644
--- a/aten/src/ATen/TensorIterator.cpp
+++ b/aten/src/ATen/TensorIterator.cpp
@@ -936,6 +936,17 @@ void TensorIteratorBase::build_binary_op(const TensorBase& out, const TensorBase
       .add_owned_input(b));
 }

+void TensorIteratorBase::shape_check_binary_op(const TensorBase& a, const TensorBase& b) {
+  build(TensorIteratorConfig()
+    .set_check_mem_overlap(true)
+    .allow_cpu_scalars(true)
+    .promote_inputs_to_common_dtype(true)
+    .cast_common_dtype_to_outputs(true)
+    .dont_allocate_output(true)
+    .add_input(a)
+    .add_input(b));
+}
+
 void TensorIteratorBase::build_borrowing_binary_op(
     const TensorBase& out, const TensorBase& a, const TensorBase& b) {
   build(BINARY_OP_CONFIG()
@@ -1420,6 +1431,8 @@ void TensorIteratorBase::build(TensorIteratorConfig& config) {
   mark_resize_outputs(config);
   // compute the result dtype and device
   compute_types(config);
+
+  if (config.dont_allocate_output_) return;
   // try fast setup output tensor, if failed, fallback to normal setup
   if (!fast_set_up(config)) {
     // compute each tensor's stride after broadcasting
diff --git a/aten/src/ATen/TensorIterator.h b/aten/src/ATen/TensorIterator.h
index e453d72d47..c8bc491b5e 100644
--- a/aten/src/ATen/TensorIterator.h
+++ b/aten/src/ATen/TensorIterator.h
@@ -424,6 +424,7 @@ public:
   void build_borrowing_binary_float_op(const TensorBase& out, const TensorBase& a, const TensorBase& b);
   TORCH_DISALLOW_TEMPORARIES(build_borrowing_binary_float_op)
   void build_binary_op(const TensorBase& out, const TensorBase& a, const TensorBase& b);
+  void shape_check_binary_op(const TensorBase& a, const TensorBase& b);
   void build_borrowing_binary_op(const TensorBase& out, const TensorBase& a, const TensorBase& b);
   TORCH_DISALLOW_TEMPORARIES(build_borrowing_binary_op)
   void build_unary_float_op(const TensorBase& out, const TensorBase& a);
@@ -719,6 +720,19 @@ public:
     return *this;
   }

+  // Sets the dont_allocate_output flag, which is false by default
+  // If true, only shape checks will run and no output will be allocated.
+  // This flag is only used by TensorIteratorBase (in structured kernels),
+  // and is not respected by TensorIterator.
+  // The only reason this flag exists is because of kernels that use TensorIteratorBase
+  // to compute intermediate results.
+  // In the meta function of an operator, we still need to run the same logic (to run shape checks),
+  // but we don't want to unnecessarily allocate an intermediate result during the meta function call.
+  TensorIteratorConfig& dont_allocate_output(const bool _dont_allocate_output) {
+    dont_allocate_output_ = _dont_allocate_output;
+    return *this;
+  }
+
   TensorIteratorConfig& resize_outputs(bool resize_outputs) {
     resize_outputs_ = resize_outputs;
     return *this;
@@ -758,6 +772,7 @@ private:
   bool promote_inputs_to_common_dtype_ = false;
   bool promote_integer_inputs_to_float_ = false;
   bool cast_common_dtype_to_outputs_ = false;
+  bool dont_allocate_output_ = false;
 };


diff --git a/aten/src/ATen/native/Loss.cpp b/aten/src/ATen/native/Loss.cpp
index 9846233440..68eb3354af 100644
--- a/aten/src/ATen/native/Loss.cpp
+++ b/aten/src/ATen/native/Loss.cpp
@@ -28,13 +28,13 @@ namespace meta {
 TORCH_META_FUNC(smooth_l1_loss)
 (const Tensor& input, const Tensor& target, const int64_t reduction, double beta) {
   TORCH_CHECK(beta >= 0, "smooth_l1_loss does not support negative values for beta.")
-  build_borrowing_binary_op(maybe_get_output(), input, target);
   if (reduction == Reduction::None) {
-    return;
+    build_borrowing_binary_op(maybe_get_output(), input, target);
+  } else {
+    TORCH_INTERNAL_ASSERT(reduction == Reduction::Mean || reduction == Reduction::Sum);
+    shape_check_binary_op(input, target);
+    at::native::resize_(maybe_get_output(), {});
   }
-
-  TORCH_INTERNAL_ASSERT(reduction == Reduction::Mean || reduction == Reduction::Sum);
-  maybe_get_output().resize_({});
 }

 } // namespace meta
```

alanwaketan replied:
@bdhirsh Thanks for the suggestion. May I ask what kinds of error are you referring to? The latest version of the pull request seems passing all tests. Does this solve the duplicated allocation problem you mentioned above?

bdhirsh replied:
oh yeah sorry maybe I was referring to an out of date comment. Yep- this should avoid the duplicated allocation

alanwaketan replied:
It looks like the above patch fail the following test case in my dev machine @bdhirsh:
```
(pytorch) jwtan@ip-10-200-23-19:/fsx/users/jwtan/work/pytorch$ cpurun python test/test_nn.py -k test_SmoothL1Loss_sum_reduction
Esss
======================================================================
ERROR: test_SmoothL1Loss_sum_reduction (__main__.TestNN)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/fsx/users/jwtan/work/pytorch/test/test_nn.py", line 12089, in <lambda>
    add(test_name, lambda self, test=test: test(self))
  File "/fsx/users/jwtan/work/pytorch/torch/testing/_internal/common_nn.py", line 6205, in __call__
    out = test_case._forward_criterion(module, input, target, extra_args=self.extra_args)
  File "/fsx/users/jwtan/work/pytorch/test/test_nn.py", line 343, in _forward_criterion
    output = criterion(input, target, *extra_args)
  File "/fsx/users/jwtan/work/pytorch/torch/nn/modules/module.py", line 1110, in _call_impl
    return forward_call(*input, **kwargs)
  File "/fsx/users/jwtan/work/pytorch/torch/nn/modules/loss.py", line 912, in forward
    return F.smooth_l1_loss(input, target, reduction=self.reduction, beta=self.beta)
  File "/fsx/users/jwtan/work/pytorch/torch/nn/functional.py", line 3116, in smooth_l1_loss
    return torch._C._nn.smooth_l1_loss(expanded_input, expanded_target, _Reduction.get_enum(reduction), beta)
RuntimeError: Expected out tensor to have dtype double, but got nullptr (uninitialized) instead

----------------------------------------------------------------------
Ran 4 tests in 0.195s

FAILED (errors=1, skipped=3)
srun: error: compile-st-c524xlarge-1: task 0: Exited with exit code 1
```

bdhirsh replied:
Okay, just shape checking and manually calling `at::native::resize_output()` isn't enough, because we also need to correctly set the dtype.

This feels worse, but I have a patch below that adds a new TIBase method, `build_borrowing_binary_op_coerce_to_scalar`, which short-circuits the set_output() path to set the size to be a scalar. I patched it on top of this PR and it passes the failing test above (for some reason I'm having issues with pytest locally and couldn't run it comprehensively across every test file though).

Maybe you can patch this in and Natalia can take a quick look at it:
```
diff --git a/aten/src/ATen/TensorIterator.cpp b/aten/src/ATen/TensorIterator.cpp
index 6fe00accf1..b2a4b0eed5 100644
--- a/aten/src/ATen/TensorIterator.cpp
+++ b/aten/src/ATen/TensorIterator.cpp
@@ -544,7 +544,13 @@ DimVector TensorIteratorBase::invert_perm(IntArrayRef input) const {
   return res;
 }

-void TensorIteratorBase::allocate_or_resize_outputs() {
+void TensorIteratorBase::allocate_or_resize_outputs(bool coerce_to_scalar) {
+  if (coerce_to_scalar) {
+    TORCH_INTERNAL_ASSERT(num_outputs_ == 1, "coerce_to_scalar is not supported for multi-output operators");
+    set_output(0, {}, {}, original_options(operands_[0]), names_);
+    return;
+  }
+
   for (const auto i : c10::irange(num_outputs_)) {
     auto& op = operands_[i];
     if (!op.tensor_base().defined() || op.will_resize) {
@@ -936,6 +942,18 @@ void TensorIteratorBase::build_binary_op(const TensorBase& out, const TensorBase
       .add_owned_input(b));
 }

+void TensorIteratorBase::build_borrowing_binary_op_coerce_to_scalar(const TensorBase& out, const TensorBase
& a, const TensorBase& b) {
+  build(TensorIteratorConfig()
+    .set_check_mem_overlap(true)
+    .allow_cpu_scalars(true)
+    .promote_inputs_to_common_dtype(true)
+    .cast_common_dtype_to_outputs(true)
+    .coerce_to_scalar(true)
+    .add_output(out)
+    .add_input(a)
+    .add_input(b));
+}
+
 void TensorIteratorBase::build_borrowing_binary_op(
     const TensorBase& out, const TensorBase& a, const TensorBase& b) {
   build(BINARY_OP_CONFIG()
@@ -1427,7 +1445,7 @@ void TensorIteratorBase::build(TensorIteratorConfig& config) {
     // re-order dimensions to improve coalescing
     reorder_dimensions();
     // allocate the output tensor if it's not provided
-    allocate_or_resize_outputs();
+    allocate_or_resize_outputs(config.coerce_to_scalar_);
     // coalesce adjacent dimensions when possible
     if (!is_meta_) coalesce_dimensions();
   }
diff --git a/aten/src/ATen/TensorIterator.h b/aten/src/ATen/TensorIterator.h
index e453d72d47..cc073b44cb 100644
--- a/aten/src/ATen/TensorIterator.h
+++ b/aten/src/ATen/TensorIterator.h
@@ -424,6 +424,7 @@ public:
   void build_borrowing_binary_float_op(const TensorBase& out, const TensorBase& a, const TensorBase& b);
   TORCH_DISALLOW_TEMPORARIES(build_borrowing_binary_float_op)
   void build_binary_op(const TensorBase& out, const TensorBase& a, const TensorBase& b);
+  void build_borrowing_binary_op_coerce_to_scalar(const TensorBase& out, const TensorBase& a, const TensorB
ase& b);
   void build_borrowing_binary_op(const TensorBase& out, const TensorBase& a, const TensorBase& b);
   TORCH_DISALLOW_TEMPORARIES(build_borrowing_binary_op)
   void build_unary_float_op(const TensorBase& out, const TensorBase& a);
@@ -445,7 +446,7 @@ protected:
   void permute_dimensions(IntArrayRef perm);
   void compute_types(const TensorIteratorConfig&);
   ScalarType compute_common_dtype();
-  void allocate_or_resize_outputs();
+  void allocate_or_resize_outputs(bool coerce_to_scalar = false);
   bool fast_set_up(const TensorIteratorConfig&);
   FastSetupType compute_fast_setup_type(const TensorIteratorConfig&);
   void compute_names(const TensorIteratorConfig&);
@@ -719,6 +720,22 @@ public:
     return *this;
   }

+  // Sets the coerce_to_scalar flag, which is false by default
+  // If true, we check that the output has scalar size, or set it if it is undefined.
+  // This flag is only used by TensorIteratorBase (in structured kernels),
+  // and is not respected by TensorIterator.
+  // The only reason this flag exists is because of kernels that use TensorIteratorBase
+  // to compute intermediate results, and then perform a reduction on the intermediate to get the result.
+  // In the meta function of an operator, we still need to run the same logic (to run shape checks),
+  // but we don't want to unnecessarily allocate an intermediate result during the meta function call.
+  //
+  // This is pretty hacky, and longterm it would be cleaner to teach TensorIterator how to deal with
+  // intermediate tensors to run the computation on, separately from the output tensor.
+  TensorIteratorConfig& coerce_to_scalar(const bool _coerce_to_scalar) {
+    coerce_to_scalar_ = _coerce_to_scalar;
+    return *this;
+  }
+
   TensorIteratorConfig& resize_outputs(bool resize_outputs) {
     resize_outputs_ = resize_outputs;
     return *this;
@@ -758,6 +775,7 @@ private:
   bool promote_inputs_to_common_dtype_ = false;
   bool promote_integer_inputs_to_float_ = false;
   bool cast_common_dtype_to_outputs_ = false;
+  bool coerce_to_scalar_ = false;
 };


diff --git a/aten/src/ATen/native/Loss.cpp b/aten/src/ATen/native/Loss.cpp
index 9846233440..494d587622 100644
--- a/aten/src/ATen/native/Loss.cpp
+++ b/aten/src/ATen/native/Loss.cpp
@@ -28,13 +28,16 @@ namespace meta {
 TORCH_META_FUNC(smooth_l1_loss)
 (const Tensor& input, const Tensor& target, const int64_t reduction, double beta) {
   TORCH_CHECK(beta >= 0, "smooth_l1_loss does not support negative values for beta.")
-  build_borrowing_binary_op(maybe_get_output(), input, target);
   if (reduction == Reduction::None) {
-    return;
+    build_borrowing_binary_op(maybe_get_output(), input, target);
+  } else {
+    TORCH_INTERNAL_ASSERT(reduction == Reduction::Mean || reduction == Reduction::Sum);
+    // We want to use TensorIteratorBase for all of the error checking in the meta kernel.
+    // However, we do NOT want to allocate the full output of the binary operator here.
+    // Why?
+    // smooth_l1_loss uses TensorIterator to construct an intermediate result, and performs a reduction to
get the output.
+    build_borrowing_binary_op_coerce_to_scalar(maybe_get_output(), input, target);
   }
-
-  TORCH_INTERNAL_ASSERT(reduction == Reduction::Mean || reduction == Reduction::Sum);
-  maybe_get_output().resize_({});
 }

 } // namespace meta
```

bdhirsh replied:
actually, I got a failure from running all of `test/test_nn.py`. Let me dig into it some more

bdhirsh replied:
Okay, this works and passes all of `test/test_nn.py`:

```
diff --git a/aten/src/ATen/TensorIterator.cpp b/aten/src/ATen/TensorIterator.cpp
index 6fe00accf1..b28e6f32dc 100644
--- a/aten/src/ATen/TensorIterator.cpp
+++ b/aten/src/ATen/TensorIterator.cpp
@@ -545,6 +545,7 @@ DimVector TensorIteratorBase::invert_perm(IntArrayRef input) const {
 }

 void TensorIteratorBase::allocate_or_resize_outputs() {
+
   for (const auto i : c10::irange(num_outputs_)) {
     auto& op = operands_[i];
     if (!op.tensor_base().defined() || op.will_resize) {
@@ -936,6 +937,14 @@ void TensorIteratorBase::build_binary_op(const TensorBase& out, const TensorBase
       .add_owned_input(b));
 }

+void TensorIteratorBase::build_borrowing_binary_op_coerce_to_scalar(const TensorBase& out, const TensorBase& a, const TensorBase& b) {
+  build(BINARY_OP_CONFIG()
+    .coerce_to_scalar(true)
+    .add_output(out)
+    .add_input(a)
+    .add_input(b));
+}
+
 void TensorIteratorBase::build_borrowing_binary_op(
     const TensorBase& out, const TensorBase& a, const TensorBase& b) {
   build(BINARY_OP_CONFIG()
@@ -1420,6 +1429,13 @@ void TensorIteratorBase::build(TensorIteratorConfig& config) {
   mark_resize_outputs(config);
   // compute the result dtype and device
   compute_types(config);
+
+  if (config.coerce_to_scalar_) {
+    TORCH_INTERNAL_ASSERT(num_outputs_ == 1, "coerce_to_scalar is not supported for multi-output operators");
+    set_output(0, {}, {}, original_options(operands_[0]), names_);
+    return;
+  }
+
   // try fast setup output tensor, if failed, fallback to normal setup
   if (!fast_set_up(config)) {
     // compute each tensor's stride after broadcasting
diff --git a/aten/src/ATen/TensorIterator.h b/aten/src/ATen/TensorIterator.h
index e453d72d47..868732fa6b 100644
--- a/aten/src/ATen/TensorIterator.h
+++ b/aten/src/ATen/TensorIterator.h
@@ -424,6 +424,7 @@ public:
   void build_borrowing_binary_float_op(const TensorBase& out, const TensorBase& a, const TensorBase& b);
   TORCH_DISALLOW_TEMPORARIES(build_borrowing_binary_float_op)
   void build_binary_op(const TensorBase& out, const TensorBase& a, const TensorBase& b);
+  void build_borrowing_binary_op_coerce_to_scalar(const TensorBase& out, const TensorBase& a, const Tensor
Base& b);
   void build_borrowing_binary_op(const TensorBase& out, const TensorBase& a, const TensorBase& b);
   TORCH_DISALLOW_TEMPORARIES(build_borrowing_binary_op)
   void build_unary_float_op(const TensorBase& out, const TensorBase& a);
@@ -719,6 +720,21 @@ public:
     return *this;
   }

+  // Sets the coerce_to_scalar flag, which is false by default
+  // If true, we check that the output has scalar size, or set it if it is undefined.
+  // This flag is only used by TensorIteratorBase (in structured kernels),
+  // and is not respected by TensorIterator.
+  // The only reason this flag exists is because of kernels that use TensorIteratorBase
+  // to compute intermediate results, and then perform a reduction on the intermediate to get the result.
+  // In the meta function of an operator, we still need to run the same logic (to run shape checks and set
 output dtype).
+  //
+  // This is pretty hacky, and longterm it would be cleaner to teach TensorIterator how to deal with
+  // intermediate tensors to run the computation on, separately from the output tensor.
+  TensorIteratorConfig& coerce_to_scalar(const bool _coerce_to_scalar) {
+    coerce_to_scalar_ = _coerce_to_scalar;
+    return *this;
+  }
+
   TensorIteratorConfig& resize_outputs(bool resize_outputs) {
     resize_outputs_ = resize_outputs;
     return *this;
@@ -758,6 +774,7 @@ private:
   bool promote_inputs_to_common_dtype_ = false;
   bool promote_integer_inputs_to_float_ = false;
   bool cast_common_dtype_to_outputs_ = false;
+  bool coerce_to_scalar_ = false;
 };


diff --git a/aten/src/ATen/native/Loss.cpp b/aten/src/ATen/native/Loss.cpp
index 9846233440..494d587622 100644
--- a/aten/src/ATen/native/Loss.cpp
+++ b/aten/src/ATen/native/Loss.cpp
@@ -28,13 +28,16 @@ namespace meta {
 TORCH_META_FUNC(smooth_l1_loss)
 (const Tensor& input, const Tensor& target, const int64_t reduction, double beta) {
   TORCH_CHECK(beta >= 0, "smooth_l1_loss does not support negative values for beta.")
-  build_borrowing_binary_op(maybe_get_output(), input, target);
   if (reduction == Reduction::None) {
-    return;
+    build_borrowing_binary_op(maybe_get_output(), input, target);
+  } else {
+    TORCH_INTERNAL_ASSERT(reduction == Reduction::Mean || reduction == Reduction::Sum);
+    // We want to use TensorIteratorBase for all of the error checking in the meta kernel.
+    // However, we do NOT want to allocate the full output of the binary operator here.
+    // Why?
+    // smooth_l1_loss uses TensorIterator to construct an intermediate result, and performs a reduction to
 get the output.
+    build_borrowing_binary_op_coerce_to_scalar(maybe_get_output(), input, target);
   }
-
-  TORCH_INTERNAL_ASSERT(reduction == Reduction::Mean || reduction == Reduction::Sum);
-  maybe_get_output().resize_({});
 }
```

alanwaketan replied:
@ngimel mind reviewing the above Brain's change to build_borrowing_binary_op_coerce_to_scalar?

Commit:
Update on "Port smooth_l1_loss to structured kernels"


Summary:
Port smooth_l1_loss to structured kernels.

Test Plan:
This commit shouldn't change the behavior. So, CI.

Differential Revision: [D31981147](https://our.internmc.facebook.com/intern/diff/D31981147)

[ghstack-poisoned]

alanwaketan commented:
@alanwaketan has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D31981147).

Commit:
Update on "Port smooth_l1_loss to structured kernels"


Summary:
Port smooth_l1_loss to structured kernels.

Test Plan:
This commit shouldn't change the behavior. So, CI.

Differential Revision: [D31981147](https://our.internmc.facebook.com/intern/diff/D31981147)

[ghstack-poisoned]

alanwaketan commented:
@alanwaketan has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D31981147).

Commit:
Update on "Port smooth_l1_loss to structured kernels"


Summary:
Port smooth_l1_loss to structured kernels.

Test Plan:
This commit shouldn't change the behavior. So, CI.

Differential Revision: [D31981147](https://our.internmc.facebook.com/intern/diff/D31981147)

[ghstack-poisoned]

alanwaketan commented:
@alanwaketan has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D31981147).

bdhirsh commented on aten/src/ATen/TensorIterator.cpp:
```diff
@@ -1420,6 +1429,13 @@ void TensorIteratorBase::build(TensorIteratorConfig& config) {
+    TORCH_INTERNAL_ASSERT(num_outputs_ == 1, "coerce_to_scalar is not supported for multi-output operators");
+    set_output(0, {}, {}, original_options(operands_[0]), names_);
+    return;
+  }
```
@ngimel wdyt. This is pretty... hacky 😬. But it solves the problem of `smooth_l1_loss` unnecessarily allocating an intermediate tensor twice. And other ops that use TensorIterator to construct an intermediate that they then perform a reduction on will have this same problem when we try to port.

It still has the downside though of needing to run `TensorIterator` twice - once in the meta function and a second time in the impl function. Long term, maybe we can teach `TensorIterator` how to handle doing the computation on an intermediate tensor separately from the output, although that would probably be a more invasive change.

ngimel commented:
Sorry for coming in late, but do you need to inherit from TensorIteratorBase, or is it just nice to have? Can you do `infer_size_dimvector` in META to get size (with proper logic for reduction and non-reduction cases), `at::result_type` to get dtype and just call set_output? And then construct `iter` in impl either with an output provided by structured kernel, or with undefined output in case of reduction? That way reduction-related logic will need to be present in both META and IMPL, but I don't think it's that bad. Let me know if you tried this already and it doesn't work.

Commit:
Update on "Port smooth_l1_loss to structured kernels"


Summary:
Port smooth_l1_loss to structured kernels.

Brian Hirsh authored the part of adding build_borrowing_binary_op_coerce_to_scalar to TensorIterator.

Test Plan:
This commit shouldn't change the behavior. So, CI.

Differential Revision: [D31981147](https://our.internmc.facebook.com/intern/diff/D31981147)

[ghstack-poisoned]

alanwaketan commented:
@alanwaketan has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D31981147).

ngimel commented:
I like this! Still extra TI construction, but looks less hacky and surprising now?

alanwaketan commented:
> I like this! Still extra TI construction, but looks less hacky and surprising now? Hopefully tests pass

Should we just land it as it is? And leave a TODO for the optimization because the PR is already hanging there for a while.

ngimel commented:
Yes, if tests pass we can land it.

Commit:
Update on "Port smooth_l1_loss to structured kernels"


Summary:
Port smooth_l1_loss to structured kernels.

Brian Hirsh authored the part of adding build_borrowing_binary_op_coerce_to_scalar to TensorIterator.

Test Plan:
This commit shouldn't change the behavior. So, CI.

Differential Revision: [D31981147](https://our.internmc.facebook.com/intern/diff/D31981147)

[ghstack-poisoned]

alanwaketan commented:
@alanwaketan has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D31981147).

alanwaketan commented:
Thanks Brian and Natalia for accepting the PR.
